storage/mysql/log_storage.go

```diff
@@ -501,6 +513,78 @@ func (t *logTreeTX) QueueLeaves(ctx context.Context, leaves []*trillian.LogLeaf,
 	return existingLeaves, nil
 }
 
+func (t *logTreeTX) AddSequencedLeaves(ctx context.Context, leaves []*trillian.LogLeaf) ([]*trillian.QueuedLogLeaf, error) {
+	res := make([]*trillian.QueuedLogLeaf, len(leaves))
+	ok := status.New(codes.OK, "OK").Proto()
+
+	// Leaves in this transaction are inserted in two tables. For each leaf, if
+	// one of the two inserts fails, we remove the side effect by rolling back to
+	// a savepoint installed before the first insert.
```

Comment:
I thought the savepoint would be updated each pass through the loop so only the ones that fail are rolled back? Is that not how it's meant to work?

Reply:
OK I now realize we're still in an all or nothing situation. Ignore that comment.

Reply:
Oh, actually we are not in all-or-nothing. Your initial interpretation was right: only the failed-to-insert entries are rolled back. This is why I update the savepoint on each loop iteration below.

Reply:
OK. It's hard to read GitHub diffs. I'll have another look.

Reply:
The reason why I create this savepoint upfront is making sure the "RELEASE SAVEPOINT ..." after the loop has something to delete and doesn't return an error.